File: M3Tools/Controls/Google/Emails/CustomEmail.vb
Imports System.Windows.Forms

Public Class CustomEmail
	Private Structure Shortcuts
		Shared Bold As String = "Bold"
		Shared Underline As String = "Underline"
		Shared Italics As String = "Italics"
	End Structure

	Public Property Subject As String
		Get
			Return ip_Subject.Text
		End Get
		Set(value As String)
			ip_Subject.Text = value
		End Set
	End Property

	Public Property Body As String()
		Get
			Return rtb_Body.Lines
		End Get
		Set(value As String())
			rtb_Body.Lines = value
		End Set
	End Property

	Public ReadOnly Property RichTextBody As String
		Get
			Return rtb_Body.Rtf
		End Get
	End Property

	Sub New()
		' This call is required by the designer.
		InitializeComponent()

		' Add any initialization after the InitializeComponent() call.
		rtb_Body.SelectionFont = fd_Font.Font
		btn_Bold.Checked = fd_Font.Font.Bold
		btn_Italics.Checked = fd_Font.Font.Italic
		btn_Underline.Checked = fd_Font.Font.Underline
	End Sub

	Private Sub ChangeFont(sender As Object, e As EventArgs) Handles btn_ChangeFont.Click
		If fd_Font.ShowDialog = DialogResult.OK Then
			rtb_Body.Font = fd_Font.Font
			btn_Bold.Checked = fd_Font.Font.Bold
			btn_Italics.Checked = fd_Font.Font.Italic
			btn_Underline.Checked = fd_Font.Font.Underline
		End If
	End Sub

	Private Sub BoldText(sender As Object, e As EventArgs) Handles btn_Bold.Click, Bold.Click
		Dim fontStyle = GetCurrentFontStyle()

		If Not rtb_Body.SelectionFont.Bold Then
			fontStyle += Drawing.FontStyle.Bold
		Else
			fontStyle -= Drawing.FontStyle.Bold
		End If

		rtb_Body.SelectionFont = New Drawing.Font(rtb_Body.Font, CType(fontStyle, Drawing.FontStyle))
	End Sub

	Private Sub BoldShortcut(sender As Object, e As EventArgs) Handles Bold.Click, Underline.Click, Italics.Click
		Dim name = CType(sender, ToolStripMenuItem).Name

		Select Case name
			Case Shortcuts.Bold
				btn_Bold.Checked = Not btn_Bold.Checked
			Case Shortcuts.Underline
				btn_Underline.Checked = Not btn_Underline.Checked
			Case Shortcuts.Italics
				btn_Italics.Checked = Not btn_Italics.Checked
		End Select
	End Sub

	Private Sub UnderlineText(sender As Object, e As EventArgs) Handles btn_Underline.Click, Underline.Click
		Dim fontStyle = GetCurrentFontStyle()

		If Not rtb_Body.SelectionFont.Underline Then
			fontStyle += Drawing.FontStyle.Underline
		Else
			fontStyle -= Drawing.FontStyle.Underline
		End If

		rtb_Body.SelectionFont = New Drawing.Font(rtb_Body.Font, CType(fontStyle, Drawing.FontStyle))

	End Sub

	Private Sub ItalicizeText(sender As Object, e As EventArgs) Handles btn_Italics.Click, Italics.Click
		Dim fontStyle = GetCurrentFontStyle()

		If Not rtb_Body.SelectionFont.Italic Then
			fontStyle += Drawing.FontStyle.Italic
		Else
			fontStyle -= Drawing.FontStyle.Italic
		End If

		rtb_Body.SelectionFont = New Drawing.Font(rtb_Body.Font, CType(fontStyle, Drawing.FontStyle))

	End Sub

	Private Function GetCurrentFontStyle() As Integer
		Dim fontStyle As Integer

		If rtb_Body.SelectionFont.Underline And rtb_Body.SelectionFont.Italic And rtb_Body.SelectionFont.Bold Then
			fontStyle = Drawing.FontStyle.Italic + Drawing.FontStyle.Underline + Drawing.FontStyle.Bold
		ElseIf rtb_Body.SelectionFont.Underline And rtb_Body.SelectionFont.Italic Then
			fontStyle = Drawing.FontStyle.Underline + Drawing.FontStyle.Italic
		ElseIf rtb_Body.SelectionFont.Underline And rtb_Body.SelectionFont.Bold Then
			fontStyle = Drawing.FontStyle.Underline + Drawing.FontStyle.Bold
		ElseIf rtb_Body.SelectionFont.Italic And rtb_Body.SelectionFont.Bold Then
			fontStyle = Drawing.FontStyle.Italic + Drawing.FontStyle.Bold
		ElseIf rtb_Body.SelectionFont.Underline Then
			fontStyle = Drawing.FontStyle.Underline
		ElseIf rtb_Body.SelectionFont.Italic Then
			fontStyle = Drawing.FontStyle.Italic
		ElseIf rtb_Body.SelectionFont.Bold Then
			fontStyle = Drawing.FontStyle.Bold
		Else
			fontStyle = 0
		End If

		Return fontStyle
	End Function
End Class

File: M3Tools/Controls/Google/Emails/CustomEmail.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;
using SPPBC.M3Tools.Types.Extensions;

// TODO: Make this have all the RTF functionality, then create a seperate view in the body selector dialog to display the RTF in HTML for 'debugging'

namespace SPPBC.M3Tools
{
	/// <summary>
	/// Holds the information for a custom email that is being sent
	/// </summary>
    public partial class CustomEmail
    {
		private readonly System.Drawing.Color _placeholderColor = System.Drawing.SystemColors.ControlDark;
		private readonly System.Drawing.Color _normalColor = System.Drawing.SystemColors.WindowText;
		private string _subjectPlaceholder = "Subject...";
		private string _bodyPlaceholder = "Body...";

        private struct Shortcuts
        {
            public static string Bold = "Bold";
            public static string Underline = "Underline";
            public static string Italics = "Italics";
        }

		/// <summary>
		/// 
		/// </summary>
		[DefaultValue("Subject...")]
		[RefreshProperties(RefreshProperties.Repaint)]
		[Category("Design")]
		public string SubejctPlaceholder
		{
			get => _subjectPlaceholder;
			set => _subjectPlaceholder = value;
		}

		/// <summary>
		/// The subject of the email
		/// </summary>
		[DefaultValue("")]
		[RefreshProperties(RefreshProperties.Repaint)]
		[Category("Email")]
		public string Subject
        {
            get
            {
				if (txt_Subject.Text == _subjectPlaceholder) return string.Empty;

                return txt_Subject.Text;
            }
			set
			{
				if (value == _bodyPlaceholder) txt_Subject.ForeColor = _placeholderColor;
				else txt_Subject.ForeColor = _normalColor;

				txt_Subject.Text = value;
			}
        }

		/// <summary>
		/// The placeholder to use for the body control
		/// </summary>
		[DefaultValue("Body...")]
		[RefreshProperties(RefreshProperties.Repaint)]
		[Category("Design")]
		public string BodyPlaceholder
		{
			get => _bodyPlaceholder;
			set => _bodyPlaceholder = value;
		}

		/// <summary>
		/// The body of the email in Rich-Text form
		/// </summary>
		[DefaultValue("")]
		[RefreshProperties(RefreshProperties.Repaint)]
		[Category("Email")]
		public string Body
        {
            get
            {
				if (rtb_Body.Text == _bodyPlaceholder) return string.Empty;
                return rtb_Body.Rtf.FromRtfToHtml();
            }
			set 
			{
				if (value == _bodyPlaceholder) rtb_Body.ForeColor = _placeholderColor;
				else rtb_Body.ForeColor = _normalColor;

				rtb_Body.Text = value;
			}
        }

		/// <summary>
		/// 
		/// </summary>
        public CustomEmail()
        {
            InitializeComponent();

			rtb_Body.KeyDown += BodyShortcutPressed;
        }

		private void BodyShortcutPressed(object sender, KeyEventArgs e)
		{
			if (!e.Control) return;

			if (e.KeyCode == Keys.B)
			{
				BoldText(sender, EventArgs.Empty);
				e.SuppressKeyPress = true;
			}
			else if (e.KeyCode == Keys.I)
			{
				ItalicizeText(sender, EventArgs.Empty);
				e.SuppressKeyPress = true;
			}
			else if (e.KeyCode == Keys.U)
			{
				UnderlineText(sender, EventArgs.Empty);
				e.SuppressKeyPress = true;
			}
		}

		private void BoldText(object sender, EventArgs e)
        {
            rtb_Body.SelectionFont = new System.Drawing.Font(rtb_Body.Font, rtb_Body.SelectionFont.Style ^ System.Drawing.FontStyle.Bold);
			btn_Bold.Checked = rtb_Body.SelectionFont.Bold;
        }

        private void UnderlineText(object sender, EventArgs e)
        {
			rtb_Body.SelectionFont = new System.Drawing.Font(rtb_Body.Font, rtb_Body.SelectionFont.Style ^ System.Drawing.FontStyle.Underline);
			btn_Underline.Checked = rtb_Body.SelectionFont.Underline;
        }

        private void ItalicizeText(object sender, EventArgs e)
        {
			rtb_Body.SelectionFont = new System.Drawing.Font(rtb_Body.Font, rtb_Body.SelectionFont.Style ^ System.Drawing.FontStyle.Italic);
			btn_Italics.Checked = rtb_Body.SelectionFont.Italic;
		}

        private void Loading(object sender, EventArgs e)
        {            
			Subject = _subjectPlaceholder;
			Body = _bodyPlaceholder;
		}

        private void TextGotFocus(object sender, EventArgs e)
        {
			Control txtBox = sender as Control;
            string placeholder = txtBox.Name == "txt_Subject" ? _subjectPlaceholder : _bodyPlaceholder;

            if (txtBox.Text != placeholder)
            {
				// No actual text present
                return;
            }

            txtBox.Text = "";
            txtBox.ForeColor = System.Drawing.SystemColors.WindowText;
        }

        private void TextLostFocus(object sender, EventArgs e)
        {
            Control txtBox = sender as Control;
            string placeholder = txtBox.Name == "txt_Subject" ? _subjectPlaceholder : _bodyPlaceholder;

            if (!string.IsNullOrEmpty(txtBox.Text))
            {
				// Actual text present
                return;
            }

			txtBox.Text = placeholder;
			txtBox.ForeColor = System.Drawing.SystemColors.ControlDark;
        }
	}
}
